File: compiler/Compiler/Compiler/Lexer/Lexer.swift
//
//  Lexer.swift
//  Compiler
//
//  Created by Yaroslav Erokhin on 20.05.2020.
//  Copyright © 2020 Yaroslav Erokhin. All rights reserved.
//

// Constants

final class Lexer {

    internal init(fileName: String? = nil, _ string: String) {
        self.fileName = fileName

        if let ascii = string.cString(using: .ascii) {
            self.stringCount = ascii.count
            self.characters = Buffer(count: stringCount, repeating: 0)
            for i in 0..<stringCount { self.characters[i] = ascii[i] }

            if string.count > 0 {
                // @Todo: bring back this error
                //            throw error(.nonASCIICharacter, start.withdrawingCharacter(), cursor)
            }

            if stringCount > 0 { self.char = characters[i] }
            else { char = C.space }
        }
        else {
            characters = Buffer(count: 0, repeating: 0)
            stringCount = 0
            char = 0
        }
    }
    
    let fileName: String?
    let characters: Buffer<CChar>
    let stringCount: Int

    // Variables
    
    var tokens: [Token] = []
    var cursor = Cursor()
    var i = 0
    var char: CChar

    let value = Buffer<CChar>()
    
    func analyze() throws -> LexerOutput {    
        while stringCount > i {
            switch char {
                
            case C.quote:
                let start = cursor
                // STRING LITERAL
                
                func isNextThreeQuotes(after n: Int = 0) -> Bool {
                    guard characters.count > i+2+n else { return false }
                    return characters[i+n] == C.quote && characters[i+1+n] == C.quote && characters[i+2+n] == C.quote
                }
                
                let isMultiline = isNextThreeQuotes()
                if isMultiline {
                    nextChar(3)
                    if !consume(C.newline) {
                        throw error(.newlineExpectedBeforeMultilineStringLiteral, cursor, cursor)
                    }
                }
                else {
                    if !nextChar() { throw error(.unexpectedEndOfFile, cursor, cursor) }
                }
                
                
                value.reset()
                while stringCount > i {
                    if char == C.newline { /* @Todo: wtf is this? */ }
                    
                    if char == C.backslash {
                        guard let next = peekNext()
                            else { throw error(.unexpectedEndOfFile, cursor, cursor) }
                        
                        switch next {
                        case C.zero:  value.append(0)
                        case C.n:  value.append(C.newline)
                        case C.r:  value.append(C.return)
                        case C.backslash: value.append(C.backslash)
                        case C.t:  value.append(C.tab)
                        case C.quote: value.append(C.quote)
                        default: throw error(.unexpectedCharacterToEscape, cursor, cursor)
                        }
                        if !nextChar() || !nextChar() { throw error(.unexpectedEndOfFile, cursor, cursor) }
                        continue
                    }
                    
                    if isMultiline {
                        if peekNext() == nil { throw error(.unexpectedEndOfFile, cursor, cursor) }
                        else if isNextThreeQuotes() {
                            throw error(.newlineExpectedAfterMultilineStringLiteral, cursor, cursor)
                        }
                        else if char == C.newline, isNextThreeQuotes(after: 1) {
                            nextChar(4)
                            if let next = peekNext(), next != C.newline, next != C.semicolon {
                                throw error(.newlineExpectedAfterMultilineStringLiteral, cursor, cursor)
                            }
                            else {
                                append(TokenLiteral(value: .string(value: C.string(from: value))), start, cursor)
                                break
                            }
                        }
                    }
                    else {
                        if consume(string: [C.quote]) {
                            append(TokenLiteral(value: .string(value: C.string(from: value))), start, cursor)
                            break
                        }
                        else if peekNext() == C.newline { throw error(.newLineInStringLiteral, cursor, cursor) }
                    }
                    
                    value.append(char)
                    if !nextChar() { throw error(.unexpectedEndOfFile, cursor, cursor) }
                }
                
            case C.semicolon, C.comma:
                // SEPARATORS
                let start = cursor
                append(Separator(value: C.string(from: [char])), start, cursor)
                
            case C.lowercaseRange, C.uppercaseRange, C.underscore, C.pound, C.accent:
                // KEYWORDS / IDENTIFIERS / DIRECTIVES / BOOL LITERALS
                
                let start = cursor
                let isDirective = consume(string: [C.pound])
                if isDirective {
                    if !nextChar() || char == 0 { throw error(.emptyDirectiveName, start, cursor) }
                    else if char == C.space { throw error(.emptyDirectiveName, start, cursor) }
                    else if !C.lowercaseRange.contains(char)
                        && !C.uppercaseRange.contains(char) && char != C.underscore {
                        throw error(.unexpectedDirectiveName, start, cursor)
                    }
                }
                
                value.reset()
                value.append(char)
                while let next = consumeNext(where: {
                    let isMatching = C.lowercaseRange.contains($0) || C.uppercaseRange.contains($0)
                    || C.numberRange.contains($0) || $0 == C.underscore || $0 == C.asterisk || $0 == C.accent
                    let isLegal = value.last != C.asterisk || value.last == C.asterisk && $0 == C.asterisk // only allowed trailing asterisks
                    return isMatching && isLegal
                }) {
                    value.append(next)
                }
                
                let isNotKeyword = value.first == C.accent && value.last == C.accent && value.count >= 3
                if isNotKeyword {
                    let copy = Buffer(value[1..<value.count-1])
                    value.reset()
                    value.append(contentsOf: copy)
                }
                
                if let idx = value.firstIndex(of: C.accent) {
                    let i = value.distance(from: value.startIndex, to: idx)
                    let c = start.advancingCharacter(by: i-1) // @Todo: weird to do this to get the right cursor
                    #if DEBUG
                    print("Unexpected char: '\(printChar)'    [@Todo: rework lexer errors 1]")
                    #endif
                    throw error(.unexpectedCharacter, c, c)
                }
                
                if value == C.void { append(TokenLiteral(value: .void), start, start) }
                else if value == C.null { append(TokenLiteral(value: .null), start, cursor) }
                else if value == C.true { append(TokenLiteral(value: .bool(value: true)), start, cursor) }
                else if value == C.false { append(TokenLiteral(value: .bool(value: false)), start, cursor) }
                else if !isNotKeyword, let keyword = Keyword(rawValue: C.string(from: value)) {
                    append(keyword, start, cursor) // @Speed: slow parsing of keywords
                }
                else if isDirective {
                    if value.isEmpty { throw error(.emptyDirectiveName, start, cursor) }
                    append(Directive(value: C.string(from: value)), start, cursor)
                }
                else {
                    if value == [C.underscore] { throw error(.invalidIdentifierUnderscore, start, cursor) }
                    append(Identifier(value: C.string(from: value)), start, cursor)
                }
                
            case C.numberRange, C.dot, C.dash:
                // NUMBER LITERALS
                
                let start = cursor
                
                // if this and the next characters are both not numbers
                if !C.numberRange.contains(char), let next = peekNext(), !C.numberRange.contains(next) {
                    fallthrough
                }
                
                value.reset()
                value.append(char)
                let numLitSymbols = [C.underscore, C.dot, C.e, C.dash]
                while let next = consumeNext(where: { C.numberRange.contains($0) || numLitSymbols.contains($0) }) {
                    if value.count >= 1 && next == C.dash && value.last != C.e {
                        throw error(.unexpectedMinusInNumberLiteral, start, cursor)
                    }
                    else if next == C.dot && value.contains(C.dot) || next == C.e && value.contains(C.e) {
                        throw error((next == C.dot ? .unexpectedDotInFloatLiteral : .unexpectedEInFloatLiteral), start, cursor)
                    }
                    if next == C.underscore { continue }
                    value.append(next)
                }
                
                if let next = peekNext() {
                    if !(C.separators + C.punctuators + C.operators).contains([next]) {
                        #if DEBUG
                        print("Unexpected char: '\(printChar)'    [@Todo: rework lexer errors 3]")
                        #endif
                        throw error(.unexpectedCharacterInNumber, cursor, cursor)
                    }
                }
                
                if value == [C.dash] || value.filter({ $0 != C.dot }).isEmpty {
                    fallthrough
                }
                else if value.contains(C.e) || value.contains(C.dot) {
                    append(TokenLiteral(value: .float(value: Float64(C.string(from: value))!)), start, cursor)
                }
                else {
                    append(TokenLiteral(value: .int(value: Int(C.string(from: value))!)), start, cursor)
                }
                
            case C.slash:
                // COMMENTS
                
                // we fallthrough to here from the case above
                // can expect ".", "e", "-" and number characters
                value.reset()
                if consume(string: [C.slash, C.slash]) {
                    guard nextChar() else { break }
                    
                    while stringCount > i {
                        if char == C.newline {
                            break
                        }
                        else {
                            value.append(char)
                            guard nextChar() else { break }
                        }
                    }
                    _ = Comment(value: C.string(from: value).trimmingCharacters(in: .whitespacesAndNewlines))
                }
                else if consume(string: C.slashAsterisk) {
                    var commentLevel = 1
                    guard nextChar() else { break }
                    
                    while stringCount > i && commentLevel > 0 {
                        if consume(string: C.slashAsterisk) {
                            commentLevel += 1
                            if commentLevel > 0 {
                                value.append(contentsOf: C.slashAsterisk)
                            }
                        }
                        else if consume(string: C.asteriskSlash) {
                            commentLevel -= 1
                            if commentLevel > 0 {
                                value.append(contentsOf: C.asteriskSlash)
                            }
                        }
                        else if char ==  C.newline {
                            value.append(C.newline)
                        }
                        else {
                            value.append(char)
                        }
                        guard nextChar() else { break }
                    }
                    _ = Comment(value: C.string(from: value).trimmingCharacters(in: .whitespacesAndNewlines))
                }
                else {
                    fallthrough
                }
                
            default:
                // PUNCTUATORS, OPERATORS
                let start = cursor
                if let value = consume(oneOf: C.punctuators) {
                    append(Punctuator(value: C.string(from: value)), start, cursor)
                }
                else if let value = consume(oneOf: C.operators) {
                    append(Operator(value: C.string(from: value)), start, cursor)
                }
                else if char == C.space || char == C.newline {
                    break
                }
                else if char != 0 {
                    #if DEBUG
                    print("Unexpected char: '\(printChar)'    [@Todo: rework lexer errors 2]")
                    #endif
                    throw error(.unexpectedCharacter, start.withdrawingCharacter(), cursor)
                }
            }
            guard char != 0 else { append(EOF(), cursor, cursor); break }
            guard nextChar() else { break }
        }
        return LexerOutput(tokens: tokens, linesProcessed: cursor.lineNumber)
    }
}

File: cppcompiler/Lexer/Lexer.cpp
//
//  Lexer.cppc
//  Compiler
//
//  Created by Ерохин Ярослав Игоревич on 23.07.2020.
//  Copyright © 2020 Yaroslav Erokhin. All rights reserved.
//

#include "Lexer.hpp"

Cursor *cursor;
int tokens_count;
Token *tokens;
int i = 0;
char character;
char* code;
unsigned long stringCount;

int value_length = 0; // value length
char* value = (char*) malloc(1000); // @Todo: dynamic buffer for longer strings

char* get_value_in_range(int min, int max) {
    char* copy = new char[max-min];
    auto pointer = (char*) &value + min;
    strcpy(copy, pointer);
    return copy;
}

char* get_value() {
    char* copy = new char[value_length];
    strcpy(copy, value);
    return copy;
}

void value_reset() {
    memset(value, 0, value_length);
    value_length = 0;
}

void value_append(char character) {
    value[value_length] = character;
    value_length += 1;
}

int index_in_value_of(char character) {
    for (int i = 0; i < value_length; i++) {  
        if (value[i] == character) {
            return i;
        }
    }
    return -1;
}

void value_append_string(char* string) {
    int length = strlen(string);
    for (int i = 0; i < length; i++) {
        value_append(string[i]);
    }
}

function<bool (char)> identifier_comparator = [](char c) {
    bool is_matching = is_in_range(c, TOKENRANGE_LOWERCASE_MIN, TOKENRANGE_LOWERCASE_MAX)
        || is_in_range(c, TOKENRANGE_UPPERCASE_MIN, TOKENRANGE_UPPERCASE_MAX)
        || is_in_range(c, TOKENRANGE_NUMBER_MIN, TOKENRANGE_NUMBER_MAX)
        || c == CHAR_UNDERSCORE
        || c == CHAR_ASTERISK
        || c == CHAR_ACCENT;
    bool is_legal = last_char_of(value) != CHAR_ASTERISK 
        || last_char_of(value) == CHAR_ASTERISK && c == CHAR_ASTERISK;
    return is_matching && is_legal;
};

function<bool (char)> number_comparator = [](char c) {
    return is_in_range(c, TOKENRANGE_NUMBER_MIN, TOKENRANGE_NUMBER_MAX)
           || c == CHAR_UNDERSCORE
           || c == CHAR_DOT
           || c == CHAR_E 
           || c == CHAR_DASH;
};

void advance(int count) {
    for (int w = 0; w < count; w++) {
        i += 1;
        if (character == CHAR_NEWLINE) {
            advance_cursor_line(cursor);
        } else {
            advance_cursor_characters(cursor, count);
        }
    }
}

bool next_char_count(int n) {
    for (int w = 0; w < n; w++) {
        advance(1);

        if (stringCount <= i) {
            return false;
        }
        character = code[i];
    }
    return true;
}

bool next_char() {
    return next_char_count(1);
}

bool consume(char query) {
    if (character == query) {
        next_char_count(0);
        return true;
    }
    return false;
}

char* consume_next(function <int (char)> compare) {
    int nextIndex = i + 1;
    if (stringCount <= nextIndex) {
        return NULL;
    }
    auto character = &code[nextIndex];
    if (compare(*character)) { // @Todo: test
        advance(1);
        return character;
    }
    return NULL;
}

bool consume_next_char(char query) {
    return consume_next([query](char character) {
        return character == query;
    }) != NULL;
}

bool consume_string(char *string) { // @Todo: test
    int index = 0;
    while (string[index] != 0 && stringCount > i + index) {
        if (code[i + index] == string[index]) {
            index += 1;
        } else {
            return false;
        }
    }
    next_char_count(index); // @Todo: test: nextChar(count-1)
    return true;
}

// goes through an array with strings of 3 characters
char* consume_one_of(char *array) {
    // @Todo: check for eof
    
    int index = 0;
    while (array[index] != 0 && stringCount > i) { // if reached the end (array[i][0] == 0)
        bool is_matching = true;
        int length = 0;
        for (int w = 0; w < 3; w++) {
            char array_value = array[index + w];
            char code_value = code[i + w];

            if (array_value != 0 && array_value != code_value) {
                is_matching = false;
                break;
            }
            if (array_value != 0 && array_value == code_value) {
                length += 1;
            }
        }

        if (is_matching) {
            auto result = new char[length];
            result[0] = code[i + 0];
            if (length > 1) {
                result[1] = code[i + 1];
            }
            if (length > 2) {
                result[2] = code[i + 2];
            }

            // 1 symbol is consumed at the end of the loop
            next_char_count(length - 1);
            return result;
        }

        index += 3;
    }
    return NULL;
}

char* peek_next_count(int n) {
    auto nextIndex = i + n;
    if (stringCount <= nextIndex) { return NULL; }
    return &code[nextIndex];
}

char* peek_next() {
    return peek_next_count(1);
}

void token_append(Token *token, Cursor *start, Cursor *end) {
    token->start = *start;
    token->end = *end;
    tokens[tokens_count] = *token;
    tokens_count += 1;
}

Token* make_token(TokenType type) {
    Token *token = (Token*) malloc(sizeof(*token));
    token->type = type;
    return token;
}

Token* make_token_separator() {
    Token *token = make_token(SEPARATOR);
    token->stringValue = get_value();
    return token;
}

bool is_next_three_quotes_after(int n) {
    if (!(stringCount > i + 2 + n)) {
        return false;
    }
    return code[i+n] == CHAR_QUOTE && code[i+1+n] == CHAR_QUOTE && code[i+2+n] == CHAR_QUOTE;
}

void fail_with_error(const char* message, Cursor *start, Cursor *end, int line_number) {
    std::cout << "error occured: " << message << "\n(context: L" << line_number << ")" << endl;
    exit(1);
}

Output* lexer_analyze(char* string) {

    // initialization
    cursor = new Cursor();
    tokens = new Token[1024];

    cursor->line_number = 1;

    code = string;
    stringCount = strlen(string);
    character = string[0];

    while (stringCount > i) {
        bool should_fallthrough = false;

        if (character == CHAR_QUOTE) {
            // STRING LITERAL
            Cursor *start = copy_cursor(cursor);

            auto is_multiline = is_next_three_quotes_after(0);
            if (is_multiline) {
                next_char_count(3);
                if (!consume(CHAR_NEWLINE)) {
                    fail_with_error("newlineExpectedBeforeMultilineStringLiteral", cursor, cursor, __LINE__);
                }
            } else {
                if (!next_char()) {
                    fail_with_error("unexpectedEndOfFile", cursor, cursor, __LINE__);
                }
            }

            value_reset();
            while (stringCount > i) {
                // if char == C.newline { /* @Todo: wtf is this? */ }

                if (character == CHAR_BACKSLASH) {
                    auto next = peek_next();
                    if (next == NULL) {
                        fail_with_error("unexpectedEndOfFile", cursor, cursor, __LINE__);
                    }

                    switch (*next) {
                    case CHAR_ZERO: { value_append(0); break; }
                    case CHAR_N: { value_append(CHAR_NEWLINE); break; }
                    case CHAR_R: { value_append(CHAR_RETURN); break; }
                    case CHAR_T: { value_append(CHAR_TAB); break; }
                    case CHAR_BACKSLASH: { value_append(CHAR_BACKSLASH); break; }
                    case CHAR_QUOTE: { value_append(CHAR_QUOTE); break; }
                    default: { fail_with_error("unexpectedCharacterToEscape", cursor, cursor, __LINE__); }
                    }

                    if (!next_char() || !next_char()) {
                        fail_with_error("unexpectedEndOfFile", cursor, cursor, __LINE__);
                    }
                    continue;
                }

                if (is_multiline) {
                    if (peek_next() == NULL) {
                        fail_with_error("unexpectedEndOfFile", cursor, cursor, __LINE__);
                    } else if (is_next_three_quotes_after(0)) {
                        fail_with_error("newlineExpectedAfterMultilineStringLiteral", cursor, cursor, __LINE__);
                    } else if (character == CHAR_NEWLINE && is_next_three_quotes_after(1)) {
                        next_char_count(4);

                        auto next = peek_next();
                        if (*next != CHAR_NEWLINE && *next != CHAR_SEMICOLON) {
                            fail_with_error("newlineExpectedAfterMultilineStringLiteral", cursor, cursor, __LINE__);
                        } else {
                            auto token = make_token(STRINGLITERAL);
                            token->stringValue = get_value();
                            token_append(token, start, cursor);
                            break;
                        }
                    }
                } else {
                    if (consume(CHAR_QUOTE)) { // @Note: if consume(string: [C.quote]) {
                        auto token = make_token(STRINGLITERAL);
                        token->stringValue = get_value();
                        token_append(token, start, cursor);
                        break;
                    } else if (*peek_next() == CHAR_NEWLINE) {
                        fail_with_error("newLineInStringLiteral", cursor, cursor, __LINE__);
                    }
                }

                value_append(character);
                if (!next_char()) {
                    fail_with_error("unexpectedEndOfFile", cursor, cursor, __LINE__);
                }
            }
        } else if (character == CHAR_SEMICOLON || character == CHAR_COMMA) {
            // SEPARATORS
            value_reset();
            value_append(character);
            auto separator_token = make_token_separator();
            token_append(separator_token, cursor, cursor);
        } else if (character == CHAR_NEWLINE || character == CHAR_SPACE) {
            // skip
        } else if (
               is_in_range(character, TOKENRANGE_LOWERCASE_MIN, TOKENRANGE_LOWERCASE_MAX)
            || is_in_range(character, TOKENRANGE_UPPERCASE_MIN, TOKENRANGE_UPPERCASE_MAX)
            || character == CHAR_UNDERSCORE
            || character == CHAR_POUND 
            || character == CHAR_ACCENT
        ) {
            // KEYWORDS / IDENTIFIERS / DIRECTIVES / BOOL LITERALS
            Cursor *start = copy_cursor(cursor);

            bool is_directive = consume(CHAR_POUND);
            if (is_directive) {
                if (!next_char() || character == 0 || character == CHAR_SPACE) {
                    fail_with_error("emptyDirectiveName", start, cursor, __LINE__);
                } else if (
                       !is_in_range(character, TOKENRANGE_LOWERCASE_MIN, TOKENRANGE_LOWERCASE_MAX)
                    && !is_in_range(character, TOKENRANGE_UPPERCASE_MIN, TOKENRANGE_UPPERCASE_MAX)
                    && character != CHAR_UNDERSCORE
                    )  {
                        fail_with_error("unexpectedDirectiveName", start, cursor, __LINE__);
                    }
            }

            value_reset();
            value_append(character);

            while (true) {
                char* next = consume_next(identifier_comparator);
                if (next == NULL) {
                    break;
                }
                value_append(*next);
            }

            bool is_not_keyword = value[0] == CHAR_ACCENT 
                && last_char_of(value) == CHAR_ACCENT
                && value_length >= 3;
            if (is_not_keyword) {
                char *copy = get_value_in_range(1, value_length-1);
                value_reset();
                value_append_string(copy);
            }

            int idx = index_in_value_of(CHAR_ACCENT);
            if (idx != -1) {
                // @Todo: add idx to cursor to 'start' to calculate the position of the unexpected symbol
                fail_with_error("unexpectedCharacter", start, start, __LINE__);
            }

            if (string_compare(value, "void")) {
                Token *token = make_token(VOIDLITERAL);
                token_append(token, start, cursor); // @Todo: looks wrong
            } else if (string_compare(value, "null\0")) {
                Token *token = make_token(NULLLITERAL);
                token_append(token, start, cursor);
            } else if (string_compare(value, "true\0")) {
                Token *token = make_token(BOOLLITERAL);
                token->boolValue = true;
                token_append(token, start, cursor);
            } else if (string_compare(value, "false\0")) {
                Token *token = make_token(BOOLLITERAL);
                token->boolValue = false;
                token_append(token, start, cursor);
            } else if (is_directive) {
                if (value == 0) {
                    fail_with_error("emptyDirectiveName", start, cursor, __LINE__);
                }
                Token *token = make_token(DIRECTIVE);
                token->stringValue = get_value();
                token_append(token, start, cursor);
            } else if (*value == CHAR_UNDERSCORE) {
                fail_with_error("invalidIdentifierUnderscore", start, cursor, __LINE__);
            } else {
                Token *token = make_token(IDENTIFIER);
                token->stringValue = get_value();
                token_append(token, start, cursor);
            }

        } else if (
            is_in_range(character, TOKENRANGE_NUMBER_MIN, TOKENRANGE_NUMBER_MAX)
            || character == CHAR_DOT
            || character == CHAR_DASH
        ) {
            // NUMBER LITERALS
            Cursor *start = copy_cursor(cursor);

            if (!is_in_range(character, TOKENRANGE_NUMBER_MIN, TOKENRANGE_NUMBER_MAX)) {
                char* next = peek_next();
                if (next == NULL || !is_in_range(*next, TOKENRANGE_NUMBER_MIN, TOKENRANGE_NUMBER_MAX)) {
                    should_fallthrough = true;
                }
            }

            if (!should_fallthrough) {
                value_reset();
                value_append(character);

                while (true) {
                    char* next = consume_next(number_comparator);
                    if (next == NULL) {
                        break;
                    }
                    if (value_length >= 1 && *next == CHAR_DASH && last_char_of(value) != CHAR_E) {
                        fail_with_error("unexpectedMinusInNumberLiteral", start, cursor, __LINE__);
                    }
                    if (*next == CHAR_DOT && string_contains(value, CHAR_DOT)
                        || *next == CHAR_E && string_contains(value, CHAR_E)) {
                        if (*next == CHAR_DOT) {
                            fail_with_error("unexpectedDotInFloatLiteral", start, cursor, __LINE__);
                        } else {
                            fail_with_error("unexpectedEInFloatLiteral", start, cursor, __LINE__);
                        }
                    }
                    if (*next == CHAR_UNDERSCORE) {
                        continue;
                    }
                    value_append(*next);
                }

                char *next = peek_next();
                if (next != NULL) {
                    // @Todo: this probably doesn't work out because 
                    // punctuators and operators are multidimentional arrays
                    if (!( string_contains((char*) separators, *next)
                        || string_contains((char*) punctuators, *next)
                        || string_contains((char*) operators, *next))
                    ) {
                        fail_with_error("unexpectedCharacterInNumber", cursor, cursor, __LINE__);
                    }
                }

                if (string_compare(value, (char*) &CHAR_DASH) || only_contains_character(value, CHAR_DOT)) {

                } else if (string_contains(value, CHAR_E) || string_contains(value, CHAR_DOT)) {
                    Token *token = make_token(FLOATLITERAL);
                    token->doubleValue = atof(value);
                    token_append(token, start, cursor);
                } else {
                    Token *token = make_token(INTLITERAL);
                    token->intValue = atoi(value);
                    token_append(token, start, cursor);
                }
            }   
        }

        if (should_fallthrough) {
            // PUNCTUATORS, OPERATORS
            Cursor *start = copy_cursor(cursor);

            auto punctuator_value = consume_one_of((char*) punctuators[0]);
            if (punctuator_value != NULL) {
                Token *token = make_token(PUNCTUATOR);
                token->stringValue = punctuator_value;
                token_append(token, start, cursor);
            } else {
                auto operator_value = consume_one_of((char*) operators[0]);
                if (operator_value != NULL) {
                    Token *token = make_token(OPERATOR);
                    token->stringValue = operator_value;
                    token_append(token, start, cursor);
                } else {
                    if (character == CHAR_SPACE || CHAR_NEWLINE) {
                        // no-op
                    } else if (character != 0) {
                        auto previous = copy_cursor(start);
                        withdraw_character(previous);
                        fail_with_error("unexpectedCharacter", previous, cursor, __LINE__);
                    }
                }
            }
        }

        if (character == 0 || !next_char()) {
            token_append(make_token(ENDOFFILE), cursor, cursor);
            break;
        }
    }

    Output *output = new Output();
    output->tokens = tokens;
    output->tokens_count = tokens_count;
    output->lines_processed = cursor->line_number;
    return output;
}
